packages/remix-server-runtime/responses.ts

```diff
@@ -1,7 +1,26 @@
-export type JsonFunction = <Data>(
+type SerializablePrimitives =
+  | string
+  | number
+  | boolean
+  | null
+  | { toJSON(): any }
+  | undefined
+  | Function
+  | symbol;
+type Serializable =
+  | SerializablePrimitives
+  | { [key: string | number | symbol]: Serializable }
+  | Serializable[];
+export type JsonFunction = <Data extends Serializable>(
   data: Data,
   init?: number | ResponseInit
-) => Response;
+) => TypedResponse<Data>;
+
+// must be a type since this is a subtype of response
+// interfaces must conform to the types they extend
+export type TypedResponse<T = any> = Response & {
```

Comment:
Doesn't really matter in this case, but I think it's a good practice to say what type it needs to extend.

Also: let's make it `unknown` as well for the same reason
https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#any

```suggestion
export type TypedResponse<T extends unknown = unknown> = Response & {
```